File: xLiAd.DiagnosticLogCenter.Agent/DiagnosticProcessors/ConsoleDiagnosticProcessor.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Reflection;
using System.Text;
using xLiAd.DiagnosticLogCenter.Abstract;

namespace xLiAd.DiagnosticLogCenter.Agent.DiagnosticProcessors
{
    public class ConsoleDiagnosticProcessor
    {
        public void BeginRequest()
        {
            StackTrace stackTrace = new StackTrace(false);
            var method = stackTrace.GetFrame(1).GetMethod();

            bool shouldRecord = FilterRule.ShouldRecord(method.DeclaringType.FullName + "." + method.Name);
            if (!shouldRecord)
            {
                GuidHolder.Holder.Value = Guid.Empty;
                return;
            }
            var guid = Guid.NewGuid();
            GuidHolder.Holder.Value = guid;
            SetTraceAndPageId();
            var log = ToLog(method);
            log.LogType = LogTypeEnum.RequestBegin;
            Helper.PostHelper.ProcessLog(log);
        }

        private void SetTraceAndPageId()
        {
            string traceId = new TracePageIdValue(DateTime.Now, DiagnosticLogConfig.Config.ClientName, DiagnosticLogConfig.Config.EnvName).ToString();
            GuidHolder.TraceIdHolder.Value = traceId;
            string pageId = traceId;
            GuidHolder.PageIdHolder.Value = pageId;

            string parentGuid = null;
            GuidHolder.ParentHolder.Value = parentGuid;

            string parentHttp = null;
            GuidHolder.ParentHttpHolder.Value = parentHttp;
        }

        private LogEntity ToLog(MethodBase method)
        {
            var path = method.DeclaringType.FullName + "." + method.Name;
            string ip = null;
            string url = null;
            LogEntity log = new LogEntity()
            {
                Message = path,
                StackTrace = url,
                MethodName = null,
                GroupGuid = GuidHolder.Holder.Value.ToString(),
                Ip = ip,
                Level = LogLeveEnum.接口日志,
                HappenTime = DateTime.Now,
                PageId = GuidHolder.PageIdHolder.Value,
                TraceId = GuidHolder.TraceIdHolder.Value,
                ParentGuid = GuidHolder.ParentHolder.Value,
                ParentHttpId = GuidHolder.ParentHttpHolder.Value
            };
            return log;
        }

        public void EndRequest()
        {
            if (GuidHolder.Holder.Value == Guid.Empty)
                return;
            StackTrace stackTrace = new StackTrace(false);
            var method = stackTrace.GetFrame(1).GetMethod();
            var log = ToLog(method);
            log.LogType = LogTypeEnum.RequestEndSuccess;
            Helper.PostHelper.ProcessLog(log);
        }

        public void RequestException(Exception exception)
        {
            if (GuidHolder.Holder.Value == Guid.Empty)
                return;
            StackTrace stackTrace = new StackTrace(false);
            var method = stackTrace.GetFrame(1).GetMethod();
            var log = ToLog(method);
            log.LogType = LogTypeEnum.RequestEndException;
            log.Message = exception.Message;
            log.StackTrace = exception.StackTrace;
            Helper.PostHelper.ProcessLog(log);
        }
    }
}

File: DiagnosticLogCenter-Java/src/main/java/com/xliad/diagnosticlogcenter/agent/diagnosticprocessors/ConsoleDiagnosticProcessor.java
package com.xliad.diagnosticlogcenter.agent.diagnosticprocessors;

import com.xliad.diagnosticlogcenter.abstracts.LogEntity;
import com.xliad.diagnosticlogcenter.abstracts.LogLevelEnum;
import com.xliad.diagnosticlogcenter.abstracts.LogTypeEnum;
import com.xliad.diagnosticlogcenter.abstracts.TracePageIdValue;
import com.xliad.diagnosticlogcenter.agent.DiagnosticLogConfig;
import com.xliad.diagnosticlogcenter.agent.FilterRule;
import com.xliad.diagnosticlogcenter.agent.GuidHolder;
import com.xliad.diagnosticlogcenter.agent.helper.PostHelper;
import org.springframework.stereotype.Component;
import java.time.LocalDateTime;
import java.util.UUID;

@Component
public class ConsoleDiagnosticProcessor {

    public void beginRequest(String methodFullName) {
        boolean shouldRecord = FilterRule.shouldRecord(methodFullName);
        if (!shouldRecord) {
            GuidHolder.setHolder(null);
            return;
        }

        String guid = UUID.randomUUID().toString();
        GuidHolder.setHolder(guid);
        setTraceAndPageId();

        LogEntity log = toLog(methodFullName);
        log.setLogType(LogTypeEnum.REQUEST_BEGIN);
        PostHelper.processLog(log);
    }

    private void setTraceAndPageId() {
        String traceId = new TracePageIdValue(
                LocalDateTime.now(),
                DiagnosticLogConfig.getConfig().getClientName(),
                DiagnosticLogConfig.getConfig().getEnvName()
        ).toString();
        GuidHolder.setTraceIdHolder(traceId);
        GuidHolder.setPageIdHolder(traceId);
        GuidHolder.setParentHolder(null);
        GuidHolder.setParentHttpHolder(null);
    }

    private LogEntity toLog(String methodFullName) {
        LogEntity log = new LogEntity();
        log.setMessage(methodFullName);
        log.setGroupGuid(GuidHolder.getHolder());
        log.setLevel(LogLevelEnum.接口日志);
        log.setHappenTime(LocalDateTime.now());
        log.setPageId(GuidHolder.getPageIdHolder());
        log.setTraceId(GuidHolder.getTraceIdHolder());
        log.setParentGuid(GuidHolder.getParentHolder());
        log.setParentHttpId(GuidHolder.getParentHttpHolder());
        return log;
    }

    public void endRequest(String methodFullName) {
        if (GuidHolder.getHolder() == null) {
            return;
        }

        LogEntity log = toLog(methodFullName);
        log.setLogType(LogTypeEnum.REQUEST_END_SUCCESS);
        PostHelper.processLog(log);
        GuidHolder.clearAll();
    }

    public void requestException(String methodFullName, Exception exception) {
        if (GuidHolder.getHolder() == null) {
            return;
        }

        LogEntity log = toLog(methodFullName);
        log.setLogType(LogTypeEnum.REQUEST_END_EXCEPTION);
        log.setMessage(exception.getMessage());
        log.setStackTrace(exception.getStackTrace().toString());
        PostHelper.processLog(log);
        GuidHolder.clearAll();
    }
}